### DISCUSSION TITLE
`atomWithHash` — unexpected `initalValue` on load

### DISCUSSION THREAD

**Discussion by alexburner** [pmndrs/jotai#2266] [Bug report] [answered]
Hi @dai-shi, this is somewhere between a bug report and a usage question... I think I'm facing the same issue as https://github.com/pmndrs/jotai/issues/739#issue-1009547194

I was trying to use `atomWithHash` to track a user's currently selected tab, like:
```tsx
import { atomWithHash } from 'jotai-location'
import React, { FC } from 'react'

const TABS = ['complete', 'running', 'error'] as const
type Tab = (typeof TABS)[number]
const isTab = (val: unknown): val is Tab => TABS.includes(val as Tab)

const INITIAL_TAB = TABS[0]

const tabAtom = atomWithHash<Tab>('tab', INITIAL_TAB, {
  serialize: (val) => val,
  deserialize: (val) => (isTab(val) ? val : INITIAL_TAB),
  setHash: 'replaceState',
})

export const Example: FC = () => {
  const tab = useAtomValue(tabAtom)
  useEffect(() => console.log(tab), [tab])
  /* expensive data fetch based on tab here */
  return <div>{tab}</div>
}
```

However, I noticed that on page load, the atom initialized twice: once with `initialValue` and once with whatever was in the URL.

I have an expensive data query associated with the currently selected tab, and landing on `#tab=running` page was fetching data for both `initialValue=success` and `currentValue=running`.

I saw that you added the `delayInit` option for this case here: https://github.com/pmndrs/jotai/pull/758/files

But it looks like `delayInit` has been removed entirely, and now the only option is `unstable_getOnInit` for `atomWithStorage`, added here: https://github.com/pmndrs/jotai/pull/1958

And I see you've also mentioned that `atomWithStorage` might be a better choice than `atomWithHash`? https://github.com/pmndrs/jotai/issues/660#issuecomment-934957986

You also mention `createHashStorage` here, but I couldn't find it in the repo: https://github.com/pmndrs/jotai/issues/739#issuecomment-929260696

So I made my own:
```ts
import { SyncStorage } from 'jotai/vanilla/utils/atomWithStorage'
import queryString from 'query-string'

type HashParams = Record<string, unknown>
const parseHash = (): HashParams => queryString.parse(location.hash.slice(1))
const stringifyHash = (params: HashParams) => `#${queryString.stringify(params)}`
const replaceHash = (hash: string) => history.replaceState(null, '', hash)

export const createHashStorage = <Value>(
  isValue: (x: unknown) => x is Value,
): SyncStorage<Value> => ({
  getItem: (key, initialValue) => {
    const params = parseHash()
    const maybeValue = params[key]
    return isValue(maybeValue) ? maybeValue : initialValue
  },
  setItem: (key, value) => {
    const params = parseHash()
    params[key] = value
    replaceHash(stringifyHash(params))
  },
  removeItem: (key) => {
    const params = parseHash()
    if (!params[key]) return
    params[key] = undefined
    replaceHash(stringifyHash(params))
  },
})
```

And am using it with `atomWithStorage` and `unstable_getOnInit`:
```tsx
const tabAtom = atomWithStorage('tab', INITIAL_TAB, createHashStorage(isTab), {
  unstable_getOnInit: true,
})
```

And, it seems to work:
- initializes with current URL hash value if present
- initializes with `initialValue` if no hash in URL

But, is this the correct solution? Or is there a way to make this work with `atomWithHash`?

**Comment by dai-shi** [comment, marked as answer]
Historically, atomWithHash was implemented with atomWithStorage (with some complexities), but it's now a separate util.
So, we should just improve atomWithHash separately.
I think it's reasonable to add `getOnInit` option to atomWithHash too.
Would you like to open a PR?

**Reply by alexburner** [reply]
That makes sense, thanks for the quick reply. And sure! I can't promise it'll be soon, but I'd be glad to try

**Reply by alexburner** [reply]
I haven't forgotten about this, but I haven't found the time to help yet

Meanwhile, I've upgraded my wrapper to also bring `initialValue` into `setItem`, so that the URL can be cleaned out if things return to their default state:

```ts
import { shallowEqual } from 'fast-equals'
import { atomWithStorage } from 'jotai/vanilla/utils'
import queryString from 'query-string'

type HashParams = Record<string, unknown>
const parseHash = (): HashParams =>
  queryString.parse(location.hash.slice(1), {
    parseNumbers: true,
    arrayFormat: 'bracket-separator',
  })
const stringifyHash = (params: HashParams) =>
  `#${queryString.stringify(params, {
    arrayFormat: 'bracket-separator',
  })}`
const replaceHash = (hash: string) => history.replaceState(null, '', hash)

export const atomWithHash = <const Value>(args: {
  key: string
  initialValue: Value
  isValue: (x: unknown) => x is Value
}) =>
  atomWithStorage(
    args.key,
    args.initialValue,
    {
      getItem: (key, initialValue) => {
        const params = parseHash()
        const maybeValue = params[key]
        return args.isValue(maybeValue) ? maybeValue : initialValue
      },
      setItem: (key, value) => {
        const params = parseHash()
        const isInitialValue = shallowEqual(value, args.initialValue)
        params[key] = isInitialValue ? undefined : value
        replaceHash(stringifyHash(params))
      },
      removeItem: (key) => {
        const params = parseHash()
        if (params[key] === undefined) return // no-op
        params[key] = undefined
        replaceHash(stringifyHash(params))
      },
    },
    {
      unstable_getOnInit: true,
    },
  )
```
